src/BinSkim.Sdk/CompilerDataLogger.cs

```diff
@@ -0,0 +1,172 @@
+
+namespace Microsoft.CodeAnalysis.IL.Sdk
+{
+    public class CompilerDataLogger : ICompilerDataLogger
+    {
+        private const string CompilerEventName = "CompilerInformation";
+
+        private readonly bool appInsightsRegistered;
+
+        private readonly string sha256;
+        private readonly string localPath;
+        private readonly string repositoryUri;
+        private readonly string pipelineName;
+
+        private static TelemetryClient s_telemetryClient;
+        private static TelemetryConfiguration s_telemetryConfiguration;
+
+        public CompilerDataLogger(IAnalysisContext analysisContext, string repositoryUri, string pipelineName)
+        {
+            try
+            {
+                string appInsightsKey = Environment.GetEnvironmentVariable("BinskimAppInsightsKey");
+                if (!string.IsNullOrEmpty(appInsightsKey) && Guid.TryParse(appInsightsKey, out _))
+                {
+                    Initialize(appInsightsKey);
+                    appInsightsRegistered = true;
+                }
+            }
+            catch (SecurityException)
+            {
+                // User does not have access to retrieve information from environment variables.
+            }
+
+            this.pipelineName = pipelineName;
+            this.repositoryUri = repositoryUri;
+            this.sha256 = analysisContext?.Hashes?.Sha256;
+            this.localPath = analysisContext?.TargetUri?.LocalPath;
+        }
+
+        public static void Initialize(string instrumentationKey)
+        {
+            s_telemetryConfiguration = new TelemetryConfiguration(instrumentationKey);
+            s_telemetryClient = new TelemetryClient(s_telemetryConfiguration);
+        }
+
+        public static void Flush()
+        {
+            s_telemetryClient?.Flush();
+        }
+
+        public void PrintHeader()
+        {
+            if (!appInsightsRegistered)
+            {
+                Console.WriteLine("RepositoryUri,PipelineName,Target,Compiler Name,Compiler BackEnd Version,Compiler FrontEnd Version,Language,Dialect,Module Name,Module Library,Hash,Error");
+            }
+        }
+
+        public void Write(string compilerData, ObjectModuleDetails omDetails)
+        {
+            string name = omDetails?.Name?.Replace(",", "_");
+            string library = omDetails?.Library?.Replace(",", ";");
+            if (appInsightsRegistered)
+            {
+                string[] compilerDataParts = compilerData.Split(',');
+
+                s_telemetryClient.TrackEvent(CompilerEventName, properties: new Dictionary<string, string>
+                {
+                    { "repositoryUri", repositoryUri ?? string.Empty },
+                    { "pipelineName", pipelineName ?? string.Empty },
+                    { "target", this.localPath },
+                    { "compilerName", compilerDataParts[0] },
+                    { "compilerBackEndVersion", compilerDataParts[1] },
```

Comment:
>compilerDataParts[1] [](http://example.com/codeflow?start=48&length=20)

look at these 'magic constants' where we have special knowledge that backend version is in the first piece. 

that tends to be pretty fragile, it might be better for this helper to break out all this data and require the caller to parse into the format.

Reply:
fixed!